src/Stratis.Bitcoin.Features.SmartContracts/Wallet/SmartContractWalletController.cs

```diff
@@ -177,68 +176,81 @@ public IActionResult GetHistory(string walletName, string address)
+                List<FlatHistory> items = accountHistory.History.OrderByDescending(o => o.Transaction.CreationTime).Where(x => x.Address.Address == address).ToList();
 
                 // Represents a sublist of transactions associated with receive addresses + a sublist of already spent transactions associated with change addresses.
                 // In effect, we filter out 'change' transactions that are not spent, as we don't want to show these in the history.
                 List<FlatHistory> history = items.Where(t => !t.Address.IsChangeAddress() || (t.Address.IsChangeAddress() && t.Transaction.IsSpent())).ToList();
 
-                foreach (FlatHistory item in history)
+                // TransactionData in history is confusingly named. A "TransactionData" actually represents an input, and the outputs that spend it are "SpendingDetails".
+                // There can be multiple "TransactionData" which have the same "SpendingDetails".
+                // For SCs we need to group spending details by their transaction ID, to get all the inputs related to the same outputs.
+                // Each group represents 1 SC transaction.
+                // Each item.Transaction in a group is an input.
+                // Each item.Transaction.SpendingDetails in the group represent the outputs, and they should all be the same so we can pick any one.
+                var scTransactions = history
+                    .Where(item => item.Transaction.SpendingDetails != null)
+                    .Where(item => item.Transaction.SpendingDetails.Payments.Any(x => x.DestinationScriptPubKey.IsSmartContractExec()))
+                    .GroupBy(item => item.Transaction.SpendingDetails.TransactionId)
+                    .Select(g => new
+                    {
+                        TransactionId = g.Key,
+                        InputAmount = g.Sum(i => i.Transaction.Amount), // Sum the inputs to the SC transaction.
+                        Outputs = g.First().Transaction.SpendingDetails.Payments, // Each item in the group will have the same outputs.
+                        OutputAmount = g.First().Transaction.SpendingDetails.Payments.Sum(o => o.Amount),
+                        BlockHeight = g.First().Transaction.SpendingDetails.BlockHeight // Each item in the group will have the same block height.
+                    })
+                    .ToList();
+
+                foreach (var scTransaction in scTransactions)
                 {
-                    bool hasSmartContract = this.receiptRepository.Retrieve(item.Transaction.Id) != null;
-                    TransactionData transaction = item.Transaction;
+                    // Consensus rules state that each transaction can have only one smart contract exec output, so FirstOrDefault is correct.
+                    PaymentDetails scPayment = scTransaction.Outputs?.FirstOrDefault(x => x.DestinationScriptPubKey.IsSmartContractExec());
```

Comment:
Because scTransations is made using `.Where(item => item.Transaction.SpendingDetails.Payments.Any(x => x.DestinationScriptPubKey.IsSmartContractExec()))`, we can use `.First()` and assume there is a SmartContractExec output?

Reply:
This would also work. I was just trying to minimise the amount of changes from the old approach.